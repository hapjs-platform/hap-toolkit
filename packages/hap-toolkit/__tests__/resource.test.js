/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const path = require('@jayfate/path')
const glob = require('glob')
const del = require('del')
const { copyApp } = require('hap-dev-utils')
const { compile } = require('../lib/commands/compile')

describe('compile a project and test resource collect', () => {
  // 资源拷贝测试，node_moduels 中的资源引用也需要拷贝到 build 目录
  // 添加 --optimize-unused-resource 参数后，未引用到的资源不应该拷贝
  it('resource collect', async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/deps-app/')
    const tempAppDir = await copyApp(testAppDir)
    const options = {
      optimizeUnusedResource: true,
      cwd: tempAppDir
    }
    const res = await compile('native', 'prod', false, options)
    const { stats } = res
    expect(stats.hasErrors()).toBeFalsy()
    const testProjectBuildPath = path.join(tempAppDir, 'build')
    const result = glob.sync('**/*', {
      cwd: testProjectBuildPath
    })

    expect(result).toMatchSnapshot('resource list')
    await del([tempAppDir], { force: true })
  }, 50000)
})

describe('css样式抽取', () => {
  // 添加 --enable-extract-css 参数后，会把ux文件中的style抽取为一个*.css.json文件
  it('Each page folder contains a *.css.json file', async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/app/')
    const tempAppDir = await copyApp(testAppDir)
    const options = {
      enableExtractCss: true,
      cwd: tempAppDir
    }
    const res = await compile('native', 'prod', false, options)
    const { stats } = res
    expect(stats.hasErrors()).toBeFalsy()
    const testProjectBuildPath = path.join(tempAppDir, 'build')
    const result = glob.sync('**/*.css.json', {
      cwd: testProjectBuildPath
    })
    const cssJsonFiles = ['CardDemo/index.css.json', 'Demo/index.css.json']
    expect(result).toEqual(expect.arrayContaining(cssJsonFiles))
    expect(result.length).toBe(2)
    expect(result).toMatchSnapshot()
    await del([tempAppDir], { force: true })
  }, 50000)
  it('make sure the rules contain meta', async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/deps-app/')
    const tempAppDir = await copyApp(testAppDir)
    const options = {
      enableExtractCss: true,
      cwd: tempAppDir
    }
    const { stats } = await compile('native', 'prod', false, options)
    expect(stats.hasErrors()).toBeFalsy()
    const testProjectBuildPath = path.join(tempAppDir, 'build')
    const result = glob.sync('**/*.css.json', {
      cwd: testProjectBuildPath
    })
    result.forEach((item) => {
      const rules = require(path.resolve(testProjectBuildPath, item)).list
      rules.forEach((rule) => {
        expect(rule).toMatchSnapshot({
          '@info': expect.any(Object)
        })
      })
    })
    await del([tempAppDir], { force: true })
  }, 50000)
})
