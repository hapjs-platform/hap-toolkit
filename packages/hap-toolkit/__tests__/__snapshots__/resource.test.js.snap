// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile a project and test resource collect resource collect: resource list 1`] = `
Array [
  "app.js",
  "assets",
  "assets/2.png",
  "assets/4.png",
  "assets/5.png",
  "assets/7.png",
  "Common",
  "Common/1.png",
  "Common/2.png",
  "Common/3.png",
  "Common/4.png",
  "Common/5.png",
  "Common/6.png",
  "Common/7.png",
  "Common/logo.png",
  "Demo",
  "Demo/index.js",
  "manifest.json",
  "node_modules",
  "node_modules/qa-test-ui",
  "node_modules/qa-test-ui/assets",
  "node_modules/qa-test-ui/assets/1.png",
  "node_modules/qa-test-ui/assets/3.png",
  "node_modules/qa-test-ui/assets/6.png",
  "node_modules/qa-test-ui/assets/7.png",
]
`;

exports[`css样式抽取 Each page folder contains a *.css.json file 1`] = `
Array [
  "CardDemo/index.css.json",
  "Demo/index.css.json",
]
`;

exports[`css样式抽取 make sure the rules contain meta 1`] = `
Object {
  ".bg-img-1": Object {
    "backgroundImage": "/node_modules/qa-test-ui/assets/3.png",
  },
  ".bg-img-2": Object {
    "backgroundImage": "/assets/4.png",
  },
  ".column": Object {
    "alignItems": "center",
    "borderBottomColor": "#cccccc",
    "borderBottomWidth": "1px",
    "borderLeftColor": "#cccccc",
    "borderLeftWidth": "1px",
    "borderRightColor": "#cccccc",
    "borderRightWidth": "1px",
    "borderStyle": "solid",
    "borderTopColor": "#cccccc",
    "borderTopWidth": "1px",
    "flexDirection": "column",
    "marginBottom": "18px",
    "marginRight": "18px",
    "width": "30%",
  },
  ".image": Object {
    "height": "200px",
    "width": "200px",
  },
  ".page": Object {
    "flexDirection": "column",
  },
  ".row": Object {
    "flexDirection": "row",
    "flexWrap": "wrap",
  },
  "@info": Any<Object>,
  "image": Object {
    "height": "200px",
    "width": "200px",
  },
  "text": Object {
    "textAlign": "center",
    "width": "100%",
  },
}
`;

exports[`css样式抽取 make sure the rules contain meta 2`] = `
Object {
  ".bg-img-1": Object {
    "backgroundImage": "/Common/3.png",
  },
  ".column": Object {
    "alignItems": "center",
    "borderBottomColor": "#cccccc",
    "borderBottomWidth": "1px",
    "borderLeftColor": "#cccccc",
    "borderLeftWidth": "1px",
    "borderRightColor": "#cccccc",
    "borderRightWidth": "1px",
    "borderStyle": "solid",
    "borderTopColor": "#cccccc",
    "borderTopWidth": "1px",
    "flexDirection": "column",
    "marginBottom": "18px",
    "marginRight": "18px",
    "width": "30%",
  },
  ".image": Object {
    "height": "200px",
    "width": "200px",
  },
  ".page": Object {
    "flexDirection": "column",
  },
  ".row": Object {
    "flexDirection": "row",
    "flexWrap": "wrap",
  },
  "@MEDIA": Array [
    Object {
      ".bg-img-2": Object {
        "backgroundImage": "/Common/4.png",
      },
      "condition": "screen and (orientation: landscape)",
    },
  ],
  "@info": Any<Object>,
  "image": Object {
    "height": "200px",
    "width": "200px",
  },
  "text": Object {
    "textAlign": "center",
    "width": "100%",
  },
}
`;
